app/src/main/java/io/edanni/movies/domain/service/MovieService.kt

```diff
@@ -0,0 +1,117 @@
+package io.edanni.movies.domain.service
+
+import io.edanni.movies.infrastructure.api.MovieApi
+import io.edanni.movies.infrastructure.api.dto.Configuration
+import io.edanni.movies.infrastructure.api.dto.Genre
+import io.edanni.movies.infrastructure.api.dto.Movie
+import io.edanni.movies.infrastructure.api.dto.MoviePage
+import io.reactivex.Observable
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.rxkotlin.Observables
+import retrofit2.Retrofit
+import javax.inject.Inject
+
+/**
+ *
+ */
+class MovieService
+@Inject constructor(retrofit: Retrofit) {
+    private val movieApi: MovieApi = retrofit.create(MovieApi::class.java)
+    private var configuration: Configuration? = null
+    private var genres: List<Genre>? = null
+
+    /**
+     * Lists upcoming movies by page.
+     */
+    fun getUpcomingMovies(page: Int): Observable<MoviePage> =
+            Observables.zip(
+                    fetchConfiguration(),
+                    fetchGenres(),
+                    movieApi.getUpcomingMovies(page),
+                    { configuration, genres: List<Genre>, moviePage: MoviePage -> Triple(configuration, genres, moviePage) })
+                    .map { (configuration, genres, movies) ->
+                        movies.copy(
+                                results = movies.results
+                                        .map { movie -> correctImagePaths(movie, configuration) }
+                                        .map { movie -> setGenres(movie, genres) }
+                        )
+                    }
+                    .observeOn(AndroidSchedulers.mainThread())
+
+    /**
+     * Searches all movies.
+     */
+    fun searchMovies(filter: String, page: Int): Observable<MoviePage> =
+            Observables.zip(
+                    fetchConfiguration(),
+                    fetchGenres(),
+                    movieApi.searchMovies(filter, page),
+                    { configuration, genres, movies -> Triple(configuration, genres, movies) })
+                    .map { (configuration, genres, movies) ->
+                        movies.copy(
+                                results = movies.results
+                                        .map { movie -> correctImagePaths(movie, configuration) }
```

Comment:
Because I need to update the image paths and genre list on the movies, and since data classes are immutable, I can only do that by returning a new list.